File: cmd/freq.go
package cmd

import (
	"flag"
	"fmt"
	"os"
	"time"

	"github.com/spf13/cobra"

	"ush/freq"
)

var freqFlags = struct {
	EncodeJSON bool
}{}

func init() {
	rootCmd.AddCommand(freqCmd)

	subcmds := []*cobra.Command{
		freqDurationCmd,
		freqExitStatusCmd,
		freqStderrCmd,
		freqStdoutCmd,
	}

	for _, cmd := range subcmds {
		flag := cmd.Flags()
		flag.BoolVar(&freqFlags.EncodeJSON, "json", false, "encode output as json")
		freqCmd.AddCommand(cmd)
	}
}

// freqCmd represents the freq subcommand.
var freqCmd = &cobra.Command{
	Use:   "freq",
	Short: "Print frequency of events from ush exec JSON output",
	Long: `Use the ush freq command to compute results from ush exec.

Examples:

	echo hello world | ush exec -- echo {.T} | ush freq exitstatus

	echo -ne 'foo\nbar\n' | ush exec -- echo {.T} | ush freq stdout

	for x in {1..3}; do echo $x; done | ush exec -p 3 -- sleep {.T} | ush freq duration 1s
`,
}

func encodeItems(items []freq.Item, err error) {
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return
	}

	if freqFlags.EncodeJSON {
		err = freq.EncodeJSON(os.Stdout, items)
	} else {
		err = freq.EncodeWide(os.Stdout, items)
	}

	if err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

var freqStdoutCmd = &cobra.Command{
	Use:   "stdout [flags] < results.json",
	Short: "Print frequency of similar stdout from ush exec JSON output",
	Run: func(cmd *cobra.Command, args []string) {
		flag.Parse()
		encodeItems(freq.Stdout(os.Stdin))
	},
}

var freqStderrCmd = &cobra.Command{
	Use:   "stderr [flags] < results.json",
	Short: "Print frequency of similar stderr from ush exec JSON output",
	Run: func(cmd *cobra.Command, args []string) {
		flag.Parse()
		encodeItems(freq.Stderr(os.Stdin))
	},
}

var freqExitStatusCmd = &cobra.Command{
	Use:   "exitstatus [flags] < results.json",
	Short: "Print frequency of similar exit status from ush exec JSON output",
	Run: func(cmd *cobra.Command, args []string) {
		flag.Parse()
		encodeItems(freq.ExitStatus(os.Stdin))
	},
}

var freqDurationCmd = &cobra.Command{
	Use:   "duration [flags] value < results.json",
	Short: "Print execution duration distribution truncated to value, e.g. 5s",
	Run: func(cmd *cobra.Command, args []string) {
		flag.Parse()

		if len(args) == 0 {
			cmd.Help()
			os.Exit(1)
		}

		d, err := time.ParseDuration(args[0])
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return
		}

		encodeItems(freq.Duration(os.Stdin, d))
	},
}

File: src/cmd/freq.rs
use std::io;

use clap::{Args, Subcommand};
use ush::freq as freq_mod;
use ush::time::parse_duration;

fn parse_duration_clap(s: &str) -> Result<std::time::Duration, String> {
    parse_duration(s).ok_or_else(|| format!("invalid duration: {s}"))
}

const EXAMPLES: &str = r"Examples:
  ush exec -- echo {} | ush freq stdout
  ush exec -- echo {} | ush freq stdout --json
  ush exec -- true | ush freq exitstatus
  ush exec -- sleep {} | ush freq duration 1s
";

#[derive(Args)]
#[command(after_help = EXAMPLES)]
pub(crate) struct FreqArgs {
    #[command(subcommand)]
    pub(crate) command: FreqCommand,
}

#[derive(Subcommand)]
pub(crate) enum FreqCommand {
    /// Print frequency of similar stdout
    #[command(after_help = EXAMPLES)]
    Stdout {
        /// Encode output as JSON
        #[arg(long)]
        json: bool,
    },
    /// Print frequency of similar stderr
    #[command(after_help = EXAMPLES)]
    Stderr {
        /// Encode output as JSON
        #[arg(long)]
        json: bool,
    },
    /// Print frequency of similar exit status
    #[command(after_help = EXAMPLES)]
    Exitstatus {
        /// Encode output as JSON
        #[arg(long)]
        json: bool,
    },
    /// Print execution duration distribution
    #[command(after_help = EXAMPLES)]
    Duration {
        /// Duration bucket size (e.g., 5s, 1m)
        #[arg(value_parser = parse_duration_clap)]
        value: std::time::Duration,
        /// Encode output as JSON
        #[arg(long)]
        json: bool,
    },
}

pub(crate) fn run(args: &FreqArgs) -> Result<(), Box<dyn std::error::Error>> {
    let stdin = io::stdin();

    match &args.command {
        FreqCommand::Stdout { json } => {
            let items = freq_mod::stdout(stdin.lock())?;
            encode_items(&items, *json)?;
        }
        FreqCommand::Stderr { json } => {
            let items = freq_mod::stderr(stdin.lock())?;
            encode_items(&items, *json)?;
        }
        FreqCommand::Exitstatus { json } => {
            let items = freq_mod::exit_status(stdin.lock())?;
            encode_items(&items, *json)?;
        }
        FreqCommand::Duration { json, value } => {
            let items = freq_mod::duration(stdin.lock(), *value)?;
            encode_items(&items, *json)?;
        }
    }

    Ok(())
}

fn encode_items(items: &[freq_mod::Item], as_json: bool) -> Result<(), Box<dyn std::error::Error>> {
    let mut stdout = io::stdout();
    if as_json {
        freq_mod::encode_json(&mut stdout, items)?;
    } else {
        freq_mod::encode_wide(&mut stdout, items)?;
    }
    Ok(())
}
